File: java/src/main/java/com/refactoring/codesmells/couplers/MessageChains.java
package com.refactoring.codesmells.couplers;

public class MessageChains {

    public static void main(String[] args) {
        Root root = new Root(new Level1(new Level2(42)));
        int value = readDeep(root);
        System.out.println("Valor profundo: " + value);
    }

    public static class Level2 {
        private int value;

        public Level2(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }
    }

    public static class Level1 {
        private Level2 next;

        public Level1(Level2 next) {
            this.next = next;
        }

        public Level2 getNext() {
            return next;
        }
    }

    public static class Root {
        private Level1 next;

        public Root(Level1 next) {
            this.next = next;
        }

        public Level1 getNext() {
            return next;
        }
    }

    public static int readDeep(Root root) {
        return root.getNext().getNext().getValue();
    }
}

File: python/src/code_smells/couplers/message_chains.py
"""
Message Chains Code Smell Example

A client asks an object for another object, which the client then asks for yet another object,
forming a chain: a.get_b().get_c().get_d()

This creates tight coupling and makes code fragile to changes in the chain.

Refactoring suggestion: Hide Delegate
"""


class Level2:
    """Deepest level in the chain"""

    def __init__(self, value: int) -> None:
        self._value = value

    def get_value(self) -> int:
        return self._value


class Level1:
    """Middle level in the chain"""

    def __init__(self, next_level: Level2) -> None:
        self._next = next_level

    def get_next(self) -> Level2:
        return self._next


class Root:
    """Root level in the chain"""

    def __init__(self, next_level: Level1) -> None:
        self._next = next_level

    def get_next(self) -> Level1:
        return self._next


def read_deep(root: Root) -> int:
    """
    Message chain: root.get_next().get_next().get_value()

    This client knows too much about the internal structure.
    Any change in the chain breaks this code.
    """
    return root.get_next().get_next().get_value()
